File: packages/api-bindings/src/subscription.ts
import {
  sendGetUsageLimitsRequest,
  sendGenerateConsoleUrlRequest,
} from "./requests.js";
import { GetUsageLimitsResponse } from "@aws/amazon-q-developer-cli-proto/fig";

export interface UsageLimits {
  currentUsage: number;
  usageLimit: number;
  overageCharges: number;
  overageEnabled: boolean;
  subscriptionTier: "free" | "pro" | "proPlus";
  resetDate: string;
}

export async function getUsageLimits(): Promise<UsageLimits> {
  const res: GetUsageLimitsResponse = await sendGetUsageLimitsRequest({});

  const usageBreakdown = res.usageBreakdown;
  const subscriptionInfo = res.subscriptionInfo;
  const overageInfo = res.overageInfo;

  let subscriptionTier: "free" | "pro" | "proPlus" = "free";
  if (subscriptionInfo) {
    switch (subscriptionInfo.tier) {
      case 0: // FREE
        subscriptionTier = "free";
        break;
      case 1: // PRO
        subscriptionTier = "pro";
        break;
      case 2: // PRO_PLUS
        subscriptionTier = "proPlus";
        break;
    }
  }

  return {
    currentUsage: usageBreakdown?.currentUsage ?? 0,
    usageLimit: usageBreakdown?.usageLimit ?? 0,
    overageCharges: usageBreakdown?.overageCharges ?? 0,
    overageEnabled: overageInfo?.enabled ?? false,
    subscriptionTier,
    resetDate: usageBreakdown?.resetDate ?? "1st of next month 12:00:00 GMT",
  };
}

export async function generateConsoleUrl(): Promise<string> {
  const response = await sendGenerateConsoleUrlRequest({});
  return response.url;
}

File: crates/fig_desktop_api/src/requests/subscription.rs
use fig_api_client::subscription::{
    SubscriptionTier as ApiTier,
    generate_console_url as generate_url,
    get_usage_limits as get_limits,
};
use fig_proto::fig::{
    GenerateConsoleUrlRequest,
    GenerateConsoleUrlResponse,
    GetUsageLimitsRequest,
    GetUsageLimitsResponse,
    OverageInfo,
    SubscriptionInfo,
    SubscriptionTier,
    UsageBreakdown,
};
use tracing::debug;

use super::{
    RequestResult,
    RequestResultImpl,
    ServerOriginatedSubMessage,
};

pub async fn get_usage_limits(_request: GetUsageLimitsRequest) -> RequestResult {
    debug!("Getting usage limits");

    match get_limits().await {
        Ok(usage_info) => {
            let usage_breakdown = UsageBreakdown {
                current_usage: usage_info.current_usage.unwrap_or(0),
                usage_limit: usage_info.usage_limit.unwrap_or(0),
                overage_charges: usage_info.overage_charges.unwrap_or(0.0),
                reset_date: usage_info
                    .reset_date_utc
                    .unwrap_or_else(|| "1st of next month 12:00:00 GMT".to_string()),
            };

            let subscription_info = SubscriptionInfo {
                tier: match usage_info.subscription_tier {
                    ApiTier::Free => SubscriptionTier::Free,
                    ApiTier::Pro => SubscriptionTier::Pro,
                    ApiTier::ProPlus => SubscriptionTier::ProPlus,
                } as i32,
            };

            let overage_info = OverageInfo {
                enabled: usage_info.overage_enabled,
            };

            Ok(
                ServerOriginatedSubMessage::GetUsageLimitsResponse(GetUsageLimitsResponse {
                    usage_breakdown: Some(usage_breakdown),
                    subscription_info: Some(subscription_info),
                    overage_info: Some(overage_info),
                })
                .into(),
            )
        },
        Err(e) => RequestResult::error(format!("Failed to get usage limits: {e}")),
    }
}

pub async fn generate_console_url(_request: GenerateConsoleUrlRequest) -> RequestResult {
    debug!("Generating console URL");

    match generate_url().await {
        Ok(url) => {
            Ok(ServerOriginatedSubMessage::GenerateConsoleUrlResponse(GenerateConsoleUrlResponse { url }).into())
        },
        Err(e) => RequestResult::error(format!("Failed to generate console URL: {e}")),
    }
}
